lib/rebaser-server/src/change_set_processor_task.rs

```diff
@@ -0,0 +1,420 @@
+    use naxum::{
+        extract::State,
+        response::{IntoResponse, Response},
+    };
+    use rebaser_core::{
+        api_types::{
+            enqueue_updates_request::EnqueueUpdatesRequest,
+            enqueue_updates_response::{
+                v1::RebaseStatus, EnqueueUpdatesResponse, EnqueueUpdatesResponseVCurrent,
+            },
+            ApiWrapper, SerializeError,
+        },
+        ContentInfo,
+    };
+    use si_data_nats::HeaderMap;
+    use telemetry::prelude::*;
+    use thiserror::Error;
+
+    use crate::{
+        middleware::{ApiTypesNegotiate, HeaderReply},
+        rebase::{perform_rebase, RebaseError},
+    };
+
+    use super::app_state::AppState;
+
+    #[remain::sorted]
+    #[derive(Debug, Error)]
+    pub(crate) enum HandlerError {
+        /// Failures related to ChangeSets
+        #[error("Change set error: {0}")]
+        ChangeSet(#[from] dal::ChangeSetError),
+        #[error("compute executor error: {0}")]
+        ComputeExecutor(#[from] dal::DedicatedExecutorError),
+        /// When failing to create a DAL context
+        #[error("error creating a dal ctx: {0}")]
+        DalTransactions(#[from] dal::TransactionsError),
+        #[error("error publishing reply: {0}")]
+        PublishReply(#[source] si_data_nats::Error),
+        /// Failures related to rebasing/updating a snapshot or change set pointer.
+        #[error("rebase error: {0}")]
+        Rebase(#[from] RebaseError),
+        #[error("error serializing: {0}")]
+        Serialize(#[from] SerializeError),
+        /// When failing to find the workspace
+        #[error("workspace error: {0}")]
+        Workspace(#[from] dal::WorkspaceError),
+        /// When failing to do an operation using the [`WorkspaceSnapshot`]
+        #[error("workspace snapshot error: {0}")]
+        WorkspaceSnapshot(#[from] dal::WorkspaceSnapshotError),
+        /// When failing to send a [`WsEvent`]
+        #[error("failed to construct ws event: {0}")]
+        WsEvent(#[from] dal::WsEventError),
+    }
+
+    type Error = HandlerError;
+
+    type Result<T> = result::Result<T, HandlerError>;
+
+    impl IntoResponse for HandlerError {
+        fn into_response(self) -> Response {
+            // TODO(fnichol): there are different responses, esp. for expected interrupted
+            error!(si.error.message = ?self, "failed to process message");
+            Response::internal_server_error()
+        }
+    }
+
+    pub(crate) async fn default(
+        State(state): State<AppState>,
+        HeaderReply(maybe_reply): HeaderReply,
+        ApiTypesNegotiate(request): ApiTypesNegotiate<EnqueueUpdatesRequest>,
+    ) -> Result<()> {
+        let AppState {
+            workspace_id,
+            change_set_id,
+            nats,
+            ctx_builder,
+            run_notify,
+        } = state;
+        let mut ctx = ctx_builder
+            .build_for_change_set_as_system(workspace_id.into(), change_set_id.into())
+            .await?;
```

Comment:
Part of me worries about how this API could be misused, but I can't think of a better way right now. We did need this though rather than making somewhat BS contexts, so thank you

Reply:
I had noticed that in many places we were using a `ctx.build_default()` (which had incorrectly set tenancy and visibility), then making other calls to mutate the `DalContext` into the right state. But yes, setting up this "god object" is becoming a little nerve-wracking.